RubberduckTests/Grammar/SelectionExtensionsTests.cs

```diff
@@ -0,0 +1,494 @@
+﻿using System.Linq;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using static Rubberduck.Parsing.Grammar.VBAParser;
+using Rubberduck.VBEditor;
+using RubberduckTests.Mocks;
+using Antlr4.Runtime.Misc;
+using System.Collections.Generic;
+
+namespace RubberduckTests.Grammar
+{
+    [TestClass]
+    public class SelectionExtensionsTests
+    {
+        public class CollectorVBAParserBaseVisitor<Result> : VBAParserBaseVisitor<IEnumerable<Result>>
+        {
+            protected override IEnumerable<Result> DefaultResult => new List<Result>();
+
+            protected override IEnumerable<Result> AggregateResult(IEnumerable<Result> firstResult, IEnumerable<Result> secondResult)
+            {
+                return firstResult.Concat(secondResult);
+            }
+        }
+
+        public class SubStmtContextElementCollectorVisitor : CollectorVBAParserBaseVisitor<SubStmtContext>
+        {
+            public override IEnumerable<SubStmtContext> VisitSubStmt([NotNull] SubStmtContext context)
+            {
+                return new List<SubStmtContext> { context };
+            }
+        }
+
+        public class IfStmtContextElementCollectorVisitor : CollectorVBAParserBaseVisitor<IfStmtContext>
+        {
+            public override IEnumerable<IfStmtContext> VisitIfStmt([NotNull] IfStmtContext context)
+            {
+                base.VisitIfStmt(context);
+                return new List<IfStmtContext> { context };
```

Comment:
If you want to get nested if statements, you have to return the result from visiting the children as well, which is the result of the call in the base class. Atm, you are determining all nested if statements and then you throw them away. 

I guess what you want to return is `VisitAllChildren(context).Concat(new List<IfStmtContext> { context}) `.

Reply:
Yikes. What a big waste! I also realized I didn't have unit tests for validating the nested statements cases (I do have two for where there are multiple but not nested). Will add. Thanks for the catch!

Reply:
FWIW: `VisitChildren` is not available to the derived class; so I went ww/ concating the base + the context and it seems to works.